smolyay/surrogate.py

```diff
@@ -470,32 +493,32 @@ def train_from_data(self, data):
         Raises
         ------
         IndexError
-            Data should be of length of the grid points x dimension.
+            Data should have size length of the grid points x dimension.
         """
         self._data = numpy.array(data, ndmin=1)
         expected_data_shape = ((len(self.grid.points), self.dimension)
                                if self.dimension > 1 else (
                                        len(self.grid.points),))
         if self._data.shape != expected_data_shape:
-            raise IndexError("Data must be of size of number"
+            raise IndexError("Data must have size of number"
                              "of grid points x dimension.")
         points, basis_functions = self.grid.points, self.grid.basis_functions
+        points = numpy.array(points)
+        num_points = len(points)
         # create basis matrix with appropriate size
         basis_matrix = numpy.zeros((self.dimension * len(points), len(points)))
         # evalaute each term
-        count = 0
-        for i, point in enumerate(points):
-            for ni in range(self.dimension):
-                for j, basis in enumerate(basis_functions):
-                    if self.dimension > 1:
-                        value = numpy.prod([f.derivative(x) if dim == ni
-                                            else f(x) for dim, (x, f)
-                                            in enumerate(zip(
-                                                 point, basis))])
-                    else:
-                        value = basis.derivative(point)
-                    basis_matrix[count, j] = value
-                count += 1
+        for ni in range(self.dimension):
+            for j, basis in enumerate(basis_functions):
+                if self.dimension > 1:
+                    value = numpy.prod(
+                            [basis[dim].derivative(points[...,dim]) 
+                             if dim == ni else basis[dim](points[...,dim]) 
+                             for dim in range(len(basis))],axis=0)
+                else:
+                    value = basis.derivative(points)
+                basis_matrix[ni + self.dimension*numpy.array(range(num_points)), 
+                              j] = value
```

Comment:
There must be a better way to write this assignment using slicing. Can you explain what you are doing?

Reply:
The way this was originally written, the row index was determined by the variable `count` which went up by 1 through the for loops for the points and the dimensions.

```
        for i, point in enumerate(points):
            for ni in range(self.dimension):
                count += 1
```

My goal was to remove the loop for the points, so I needed a way to figure out the value of count so that all the elements in `value` get assigned to the right place in `basis_matrix`.

The value of `count` can be calculated at every step by `ni + self.dimension*i`. Removing that first for loop means that you're calculating all values that go in indexes `ni + self.dimension*i for i in range(0, len(points))`. What's currently written is a more compact way of getting all those indexes. These indexes aren't next to each other (for example when ni = 1 and self.dimension = 3 the indexes are 1, 4, 7, 10, 13, ... etc) which makes it more complicated.

Reply:
I see, so if you want to assign to indexes starting from `ni` in steps of `self.dimension`, you can use standard Python slicing:
```
basis_matrix[ni::self.dimension, j] = value
```

Reply:
It is done.